src/manifolds/Decorator.jl

```diff
@@ -0,0 +1,560 @@
+#
+# General functions for decorators
+#
+"""
+    AbstractDecoratorManifold<:Manifold
+
+"""
+abstract type AbstractDecoratorManifold<:Manifold end
+"""
+    is_default_decorator(M)
+
+For any manifold that [`is_decorator_manifold`](@ref) it might be the default
+decorator. This function dispatches onto these checks for specific decorators
+"""
+is_default_decorator(M::Manifold) = false
+
+"""
+    is_decorator_transparent(M, f)
+
+Given a [`Manifold`](@ref) `M` and a function `f`, indicate, whether a
+[`is_decorator_manifold`](@ref) acts transparent for `f`. This means, it
+just passes through down to the internally stored manifold.
+Only decorator manifolds can be transparent and their default is, to be transparent.
+A function that is affected by the decorator hence has to set this to `Val(false)`
+actively.
+If a decorator manifold is not in general transparent, it might still pass down
+for the case that a decorator is the default decorator, see [`is_default_decorator`](@ref).
+"""
+is_decorator_transparent(M::Manifold, f) = is_decorator_transparent(M::Manifold, f, is_decorator_manifold(M))
+is_decorator_transparent(M::Manifold, f, ::Val{true}) = true
+is_decorator_transparent(M::Manifold, f, ::Val{false}) = false
+
+
+function manifold_function_not_implemented_message(M,f,x...)
+    s = join(string.([x...]),", "," and ")
+    a = length(x) > 1 ? "arguments" : "argument"
+    m = length(x) > 0 ? " for $(a) $(s)." : "."
+    return "$(f) not implemented on $(M)$(m)"
+end
+
+_acts_transparent(M::Manifold, f) = is_default_decorator(M) || is_decorator_transparent(M,f)
```

Comment:
```suggestion
_acts_transparently(M::Manifold, f) = is_default_decorator(M) || is_decorator_transparent(M,f)
```

Reply:
For this one I have to `or` to `Val`-functions then in order to be efficient?

Reply:
You can for example make a function
```julia
val_or(::Val{T1}, ::Val{T2}) where {T1,T2} = Val(T1 || T2)
```
and use it there.